// cubitt-commands/dist/CommandFactory.js
"use strict";
var Common = require("cubitt-common");
var CommandType_1 = require("./CommandType");
var AddConnectorCommand_1 = require("./commands/AddConnectorCommand");
var AddEdgeCommand_1 = require("./commands/AddEdgeCommand");
var AddModelCommand_1 = require("./commands/AddModelCommand");
var AddNodeCommand_1 = require("./commands/AddNodeCommand");
var DeleteConnectorCommand_1 = require("./commands/DeleteConnectorCommand");
var DeleteEdgeCommand_1 = require("./commands/DeleteEdgeCommand");
var DeleteModelCommand_1 = require("./commands/DeleteModelCommand");
var DeleteNodeCommand_1 = require("./commands/DeleteNodeCommand");
var DeleteConnectorPropertyCommand_1 = require("./commands/DeleteConnectorPropertyCommand");
var DeleteEdgePropertyCommand_1 = require("./commands/DeleteEdgePropertyCommand");
var DeleteModelPropertyCommand_1 = require("./commands/DeleteModelPropertyCommand");
var DeleteNodePropertyCommand_1 = require("./commands/DeleteNodePropertyCommand");
var SetConnectorPropertyCommand_1 = require("./commands/SetConnectorPropertyCommand");
var SetEdgePropertyCommand_1 = require("./commands/SetEdgePropertyCommand");
var SetModelPropertyCommand_1 = require("./commands/SetModelPropertyCommand");
var SetNodePropertyCommand_1 = require("./commands/SetNodePropertyCommand");
/**
 * Factory for creation of Commands from JSON objects.
 */
var CommandFactory = (function () {
    function CommandFactory() {
    }
    /**
     * Attempts to create a Command from a JSON object.
     *
     * @param jsonObject JSON Object representing the Command.
     */
    CommandFactory.parse = function (jsonObject) {
        if (typeof (jsonObject) === "undefined" || jsonObject === null) {
            throw new Error("COMMAND ERROR: Empty object cannot be parsed.");
        }
        var typeString = jsonObject["type"];
        if (typeof (typeString) === "undefined" || typeString === null) {
            throw new Error("COMMAND ERROR: Object has no 'type' property.");
        }
        var type = CommandFactory.getCommandType(typeString);
        var keys = {
            id: ValidationType.Guid,
            requestId: ValidationType.Guid,
            sessionId: ValidationType.Guid
        };
        switch (type) {
            case CommandType_1.CommandType.AddConnector:
                return CommandFactory.parseAddConnectorCommand(jsonObject, keys);
            case CommandType_1.CommandType.AddEdge:
                return CommandFactory.parseAddEdgeCommand(jsonObject, keys);
            case CommandType_1.CommandType.AddModel:
                return CommandFactory.parseAddModelCommand(jsonObject, keys);
            case CommandType_1.CommandType.AddNode:
                return CommandFactory.parseAddNodeCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteConnector:
                return CommandFactory.parseDeleteConnectorCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteEdge:
                return CommandFactory.parseDeleteEdgeCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteModel:
                return CommandFactory.parseDeleteModelCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteNode:
                return CommandFactory.parseDeleteNodeCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteConnectorProperty:
                return CommandFactory.parseDeleteConnectorPropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteEdgeProperty:
                return CommandFactory.parseDeleteEdgePropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteModelProperty:
                return CommandFactory.parseDeleteModelPropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.DeleteNodeProperty:
                return CommandFactory.parseDeleteNodePropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.SetConnectorProperty:
                return CommandFactory.parseSetConnectorPropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.SetEdgeProperty:
                return CommandFactory.parseSetEdgePropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.SetModelProperty:
                return CommandFactory.parseSetModelPropertyCommand(jsonObject, keys);
            case CommandType_1.CommandType.SetNodeProperty:
                return CommandFactory.parseSetNodePropertyCommand(jsonObject, keys);
            /* istanbul ignore next */
            default:
                throw new Error("COMMAND ERROR: CommandType '" + type.toString() + "' can not be parsed.");
        }
    };
    /**
     * Validates the keys of the passed objects against the passed ValidationTypes.
     *
     * @param object Object that should be validated.
     * @param keysWithType Dictionary of the keys that will be validated, with as value the type of validation.
     */
    CommandFactory.validateObject = function (object, keysWithType) {
        for (var key in keysWithType) {
            var type = keysWithType[key];
            var val = object[key];
            if (typeof (val) === "undefined") {
                throw new Error("COMMAND ERROR: Attribute '" + key.toString() + "' is mising.");
            }
            switch (type) {
                case ValidationType.Guid:
                    var guid = Common.Guid.parse(val);
                    if (guid === null) {
                        throw new Error("COMMAND ERROR: Attribute '" + key.toString() + "' is not a valid GUID.");
                    }
                    break;
                case ValidationType.String:
                    if (val.toString().trim().length === 0) {
                        throw new Error("COMMAND ERROR: Attribute '" + key.toString() + "' is missing or empty.");
                    }
                    break;
                case ValidationType.Any:
                    break;
                /* istanbul ignore next */
                default:
                    throw new Error("COMMAND ERROR: ValidationType " + type.toString() + " is not valid.");
            }
        }
    };
    /**
     * Parses a properties object and returns a valid properties object.
     *
     * @param properties.
     */
    CommandFactory.parseProperties = function (properties) {
        if (typeof (properties) === "undefined" || properties === null) {
            return {};
        }
        if (typeof (properties["type"]) !== "undefined") {
            delete properties["type"];
        }
        return properties;
    };
    /**
     * Attempts to create a CommandType from a string.
     *
     * @param type string containing a CommandType text.
     */
    CommandFactory.getCommandType = function (type) {
        switch (type) {
            case "AddConnectorCommand":
                return CommandType_1.CommandType.AddConnector;
            case "AddEdgeCommand":
                return CommandType_1.CommandType.AddEdge;
            case "AddModelCommand":
                return CommandType_1.CommandType.AddModel;
            case "AddNodeCommand":
                return CommandType_1.CommandType.AddNode;
            case "DeleteConnectorCommand":
                return CommandType_1.CommandType.DeleteConnector;
            case "DeleteEdgeCommand":
                return CommandType_1.CommandType.DeleteEdge;
            case "DeleteModelCommand":
                return CommandType_1.CommandType.DeleteModel;
            case "DeleteNodeCommand":
                return CommandType_1.CommandType.DeleteNode;
            case "SetConnectorPropertyCommand":
                return CommandType_1.CommandType.SetConnectorProperty;
            case "SetEdgePropertyCommand":
                return CommandType_1.CommandType.SetEdgeProperty;
            case "SetModelPropertyCommand":
                return CommandType_1.CommandType.SetModelProperty;
            case "SetNodePropertyCommand":
                return CommandType_1.CommandType.SetNodeProperty;
            case "DeleteConnectorPropertyCommand":
                return CommandType_1.CommandType.DeleteConnectorProperty;
            case "DeleteEdgePropertyCommand":
                return CommandType_1.CommandType.DeleteEdgeProperty;
            case "DeleteModelPropertyCommand":
                return CommandType_1.CommandType.DeleteModelProperty;
            case "DeleteNodePropertyCommand":
                return CommandType_1.CommandType.DeleteNodeProperty;
            /* istanbul ignore next */
            default:
                throw new Error("COMMAND ERROR: CommandType " + type.toString() + " is not valid.");
        }
    };
    /**
     * Attempts to create an AddConnectorCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseAddConnectorCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["elementType"] = ValidationType.String;
        keys["nodeId"] = ValidationType.Guid;
        var properties = CommandFactory.parseProperties(jsonObject["elementProperties"]);
        CommandFactory.validateObject(jsonObject, keys);
        return new AddConnectorCommand_1.AddConnectorCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["elementType"].toString(), properties, Common.Guid.parse(jsonObject["nodeId"]));
    };
    /**
     * Attempts to create an AddEdgeCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseAddEdgeCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["elementType"] = ValidationType.String;
        keys["modelId"] = ValidationType.Guid;
        keys["startConnectorId"] = ValidationType.Guid;
        keys["endConnectorId"] = ValidationType.Guid;
        var properties = CommandFactory.parseProperties(jsonObject["elementProperties"]);
        CommandFactory.validateObject(jsonObject, keys);
        return new AddEdgeCommand_1.AddEdgeCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["elementType"].toString(), properties, Common.Guid.parse(jsonObject["modelId"]), Common.Guid.parse(jsonObject["startConnectorId"]), Common.Guid.parse(jsonObject["endConnectorId"]));
    };
    /**
     * Attempts to create an AddModelCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseAddModelCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["elementType"] = ValidationType.String;
        var properties = CommandFactory.parseProperties(jsonObject["elementProperties"]);
        CommandFactory.validateObject(jsonObject, keys);
        return new AddModelCommand_1.AddModelCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["elementType"].toString(), properties);
    };
    /**
     * Attempts to create an AddNodeCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseAddNodeCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["elementType"] = ValidationType.String;
        keys["modelId"] = ValidationType.Guid;
        var properties = CommandFactory.parseProperties(jsonObject["elementProperties"]);
        CommandFactory.validateObject(jsonObject, keys);
        return new AddNodeCommand_1.AddNodeCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["elementType"].toString(), properties, Common.Guid.parse(jsonObject["modelId"]));
    };
    /**
     * Attempts to create an DeleteConnectorCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteConnectorCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteConnectorCommand_1.DeleteConnectorCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]));
    };
    /**
     * Attempts to create an DeleteEdgeCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteEdgeCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteEdgeCommand_1.DeleteEdgeCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]));
    };
    /**
     * Attempts to create an DeleteModelCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteModelCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteModelCommand_1.DeleteModelCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]));
    };
    /**
     * Attempts to create an DeleteNodeCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteNodeCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteNodeCommand_1.DeleteNodeCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]));
    };
    /**
     * Attempts to create a DeleteConnectorPropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteConnectorPropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteConnectorPropertyCommand_1.DeleteConnectorPropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString());
    };
    /**
     * Attempts to create a DeleteEdgePropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteEdgePropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteEdgePropertyCommand_1.DeleteEdgePropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString());
    };
    /**
     * Attempts to create a DeleteModelPropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteModelPropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteModelPropertyCommand_1.DeleteModelPropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString());
    };
    /**
     * Attempts to create a DeleteNodePropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseDeleteNodePropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        CommandFactory.validateObject(jsonObject, keys);
        return new DeleteNodePropertyCommand_1.DeleteNodePropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString());
    };
    /**
     * Attempts to create a SetConnectorPropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseSetConnectorPropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        keys["propertyValue"] = ValidationType.Any;
        CommandFactory.validateObject(jsonObject, keys);
        return new SetConnectorPropertyCommand_1.SetConnectorPropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString(), jsonObject["propertyValue"]);
    };
    /**
     * Attempts to create a SetEdgePropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseSetEdgePropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        keys["propertyValue"] = ValidationType.Any;
        CommandFactory.validateObject(jsonObject, keys);
        return new SetEdgePropertyCommand_1.SetEdgePropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString(), jsonObject["propertyValue"]);
    };
    /**
     * Attempts to create a SetModelPropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseSetModelPropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        keys["propertyValue"] = ValidationType.Any;
        CommandFactory.validateObject(jsonObject, keys);
        return new SetModelPropertyCommand_1.SetModelPropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString(), jsonObject["propertyValue"]);
    };
    /**
     * Attempts to create a SetNodePropertyCommand.
     *
     * @param jsonObject JSON Object representing the Command.
     * @param keys Keys of the jsonObject that will be validated against the matched ValidationTypes.
     */
    CommandFactory.parseSetNodePropertyCommand = function (jsonObject, keys) {
        keys["elementId"] = ValidationType.Guid;
        keys["propertyName"] = ValidationType.String;
        keys["propertyValue"] = ValidationType.Any;
        CommandFactory.validateObject(jsonObject, keys);
        return new SetNodePropertyCommand_1.SetNodePropertyCommand(Common.Guid.parse(jsonObject["id"]), Common.Guid.parse(jsonObject["requestId"]), Common.Guid.parse(jsonObject["sessionId"]), Common.Guid.parse(jsonObject["elementId"]), jsonObject["propertyName"].toString(), jsonObject["propertyValue"]);
    };
    return CommandFactory;
}());
exports.CommandFactory = CommandFactory;
/**
 * An enum for all types of attributes that can be validated.
 */
(function (ValidationType) {
    ValidationType[ValidationType["Guid"] = 0] = "Guid";
    ValidationType[ValidationType["String"] = 1] = "String";
    ValidationType[ValidationType["Properties"] = 2] = "Properties";
    ValidationType[ValidationType["Any"] = 3] = "Any";
})(exports.ValidationType || (exports.ValidationType = {}));
var ValidationType = exports.ValidationType;
