// cubitt-graph/dist/Project.js
"use strict";
var Graph_1 = require("./Graph");
var Project = (function () {
    function Project() {
        this.graph = new Graph_1.Graph();
    }
    /**
     * @inheritdoc
     */
    Project.prototype.addNode = function (id, type, modelId, properties) {
        if (properties === void 0) { properties = {}; }
        this.graph.addNode(id, type, modelId, properties);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.addEdge = function (id, type, modelId, startConnectorId, endConnectorId, properties) {
        this.graph.addEdge(id, type, modelId, startConnectorId, endConnectorId, properties);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.addConnector = function (id, type, nodeId, properties) {
        if (properties === void 0) { properties = {}; }
        this.graph.addConnector(id, type, nodeId, properties);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.addModel = function (id, type, properties, parentId) {
        if (properties === void 0) { properties = {}; }
        this.graph.addModel(id, type, properties, parentId);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.setProperty = function (id, name, value) {
        this.graph.setProperty(id, name, value);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.deleteNode = function (id) {
        this.graph.deleteNode(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.deleteEdge = function (id) {
        this.graph.deleteEdge(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.deleteConnector = function (id) {
        this.graph.deleteConnector(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.deleteModel = function (id) {
        this.graph.deleteModel(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.deleteProperty = function (id, key) {
        this.graph.deleteProperty(id, key);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.serialize = function () {
        return this.graph.serialize();
    };
    /**
     * @inheritdoc
     */
    Project.prototype.deserialize = function (jsonObject) {
        return this.graph.deserialize(jsonObject);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.hasElement = function (id) {
        return this.graph.hasElement(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.hasModel = function (id) {
        return this.graph.hasModel(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.hasNode = function (id) {
        return this.graph.hasNode(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.hasConnector = function (id) {
        return this.graph.hasConnector(id);
    };
    /**
     * @inheritdoc
     */
    Project.prototype.hasEdge = function (id) {
        return this.graph.hasEdge(id);
    };
    return Project;
}());
exports.Project = Project;
